packages/shorebird_cli/lib/src/artifact_builder/flutter_build_log_updater.dart

```diff
@@ -0,0 +1,108 @@
+import 'dart:collection';
+
+import 'package:shorebird_cli/src/platform.dart';
+
+/// A callback that is called when a new step is identified in the build
+typedef BuildStepCallback = void Function(LogUpdaterStep step);
+
+/// This enum describes the different steps that the log updater
+/// identifies in the build process.
+///
+/// Not all of the steps apply to all platforms or to the situation of the
+/// user's build (e.g. [downloadingGradleW] only applies to Android projects
+/// in the first build).
+///
+/// Refer to the comment of each step for more information.
+enum LogUpdaterStep {
+  /// The build has just started.
+  initial,
+
+  /// Generic step called when the build is happening, but
+  /// no specifc, or outstanding, step has been identified.
+  building,
+
+  /// The project does not have a gradle wrapper binary and it is being
+  /// downloaded. (Android only)
+  downloadingGradleW,
+
+  /// Flutter is checking and preparing the Android SDK. (Android only)
+  preparingAndroidSDK,
```

Comment:
I wonder if this should be a super enum and store the message and parsing regexp as variables. 🤷 

Or I guess alternatively it could just be a list of objects which implement those two methods.  This design is fine too, mostly musing.